src/main/java/net/wimpi/modbus/io/ModbusTCPTransport.java

```diff
@@ -95,16 +97,30 @@ public void close() throws IOException {
     @Override
     public void writeMessage(ModbusMessage msg) throws ModbusIOException {
         try {
-            msg.writeTo(m_Output);
+            byte[] serializedMessage = serialize(msg);
+            m_Output.write(serializedMessage);
             m_Output.flush();
             // write more sophisticated exception handling
-        } catch (Exception ex) {
+        } catch (IOException ex) {
             final String errMsg = "I/O exception - failed to write";
             logger.error(errMsg, ex);
             throw new ModbusIOException(String.format("%s: %s", errMsg, ex.getMessage()));
         }
     }// write
 
+    private byte[] serialize(ModbusMessage msg) throws IOException {
+        try (
+                ByteArrayOutputStream buffer = new ByteArrayOutputStream();
+                DataOutputStream bufferOutputStream = new DataOutputStream(buffer)
+        ) {
+            msg.writeTo(bufferOutputStream);
+            bufferOutputStream.flush();
+            return buffer.toByteArray();
+        } catch (RuntimeException e) {
```

Comment:
Why not catch all exceptions?

Reply:
The IOException (which is quite unlikely to occur with a ByteArrayOutputStream anyway) is already handled one level above. And there are no other checked exceptions thrown by this code. So I don't see a reason to catch more than I need.

Reply:
Ok